File: packages/modules/Bluetooth/system/bta/le_audio/content_control_id_keeper.cc
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "content_control_id_keeper.h"

#include <bitset>
#include <map>

#include "le_audio_types.h"
#include "osi/include/log.h"

namespace {

using le_audio::types::LeAudioContextType;

}  // namespace

namespace le_audio {

struct ccid_keeper {
 public:
  ccid_keeper() {}

  ~ccid_keeper() {}

  void SetCcid(uint16_t context_type, int ccid) {
    LOG_DEBUG("Ccid: %d, context type %d", ccid, context_type);

    std::bitset<16> test{context_type};
    if (test.count() > 1 ||
        context_type >=
            static_cast<std::underlying_type<LeAudioContextType>::type>(
                LeAudioContextType::RFU)) {
      LOG_ERROR("Unknownd context type %d", context_type);
      return;
    }

    auto ctx_type = static_cast<LeAudioContextType>(context_type);
    ccids_.insert_or_assign(ctx_type, ccid);
  }

  int GetCcid(uint16_t context_type) const {
    std::bitset<16> test{context_type};
    if (test.count() > 1 ||
        context_type >=
            static_cast<std::underlying_type<LeAudioContextType>::type>(
                LeAudioContextType::RFU)) {
      LOG_ERROR("Unknownd context type %d", context_type);
      return -1;
    }

    auto ctx_type = static_cast<LeAudioContextType>(context_type);

    if (ccids_.count(ctx_type) == 0) {
      return -1;
    }

    return ccids_.at(ctx_type);
  }

 private:
  /* Ccid informations */
  std::map<LeAudioContextType /* context */, int /*ccid */> ccids_;
};

struct ContentControlIdKeeper::impl {
  impl(const ContentControlIdKeeper& ccid_keeper) : ccid_keeper_(ccid_keeper) {}

  void Start() {
    LOG_ASSERT(!ccid_keeper_impl_);
    ccid_keeper_impl_ = std::make_unique<ccid_keeper>();
  }

  void Stop() {
    LOG_ASSERT(ccid_keeper_impl_);
    ccid_keeper_impl_.reset();
  }

  bool IsRunning() { return ccid_keeper_impl_ ? true : false; }

  const ContentControlIdKeeper& ccid_keeper_;
  std::unique_ptr<ccid_keeper> ccid_keeper_impl_;
};

ContentControlIdKeeper::ContentControlIdKeeper()
    : pimpl_(std::make_unique<impl>(*this)) {}

void ContentControlIdKeeper::Start() {
  if (!pimpl_->IsRunning()) pimpl_->Start();
}

void ContentControlIdKeeper::Stop() {
  if (pimpl_->IsRunning()) pimpl_->Stop();
}

int ContentControlIdKeeper::GetCcid(uint16_t context_type) const {
  if (!pimpl_->IsRunning()) {
    return -1;
  }

  return pimpl_->ccid_keeper_impl_->GetCcid(context_type);
}

void ContentControlIdKeeper::SetCcid(uint16_t context_type, int ccid) {
  if (pimpl_->IsRunning()) {
    pimpl_->ccid_keeper_impl_->SetCcid(context_type, ccid);
  }
}

}  // namespace le_audio

File: packages/modules/Bluetooth/android/app/src/com/android/bluetooth/le_audio/ContentControlIdKeeper.java
/*
 * Copyright 2021 HIMSA II K/S - www.himsa.com.
 * Represented by EHIMA - www.ehima.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.le_audio;

import android.bluetooth.BluetoothLeAudio;
import android.os.ParcelUuid;
import android.util.Pair;

import com.android.bluetooth.btservice.ServiceFactory;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.SortedSet;
import java.util.TreeSet;

/**
 * This class keeps Content Control Ids for LE Audio profiles.
 */
public class ContentControlIdKeeper {

    public static final int CCID_INVALID = 0;
    public static final int CCID_MIN = 0x01;
    public static final int CCID_MAX = 0xFF;

    private static SortedSet<Integer> sAssignedCcidList = new TreeSet();
    private static HashMap<ParcelUuid, Pair<Integer, Integer>> sUserMap = new HashMap();
    private static ServiceFactory sServiceFactory = null;

    /**
     * Functions is used to acquire Content Control ID (Ccid). Ccid is connected
     * with a context type  and the user uuid. In most of cases user uuid is the GATT service
     * UUID which makes use of Ccid
     *
     * @param userUuid user identifier (GATT service)
     * @param contextType the context types as defined in {@link BluetoothLeAudio}
     * @return ccid to be used in the Gatt service Ccid characteristic.
    */
    public static synchronized int acquireCcid(ParcelUuid userUuid, int contextType) {
        int ccid = CCID_INVALID;

        if (sAssignedCcidList.size() == 0) {
            ccid = CCID_MIN;
        } else if (sAssignedCcidList.last() < CCID_MAX) {
            ccid = sAssignedCcidList.last() + 1;
        } else if (sAssignedCcidList.first() > CCID_MIN) {
            ccid = sAssignedCcidList.first() - 1;
        } else {
            int first_ccid_avail = sAssignedCcidList.first() + 1;
            while (first_ccid_avail < CCID_MAX - 1) {
                if (!sAssignedCcidList.contains(first_ccid_avail)) {
                    ccid = first_ccid_avail;
                    break;
                }
                first_ccid_avail++;
            }
        }

        if (ccid != CCID_INVALID)  {
            sAssignedCcidList.add(ccid);
            sUserMap.put(userUuid, new Pair(ccid, contextType));

            if (sServiceFactory == null) {
                sServiceFactory = new ServiceFactory();
            }
            /* Notify LeAudioService about new ccid  */
            LeAudioService service = sServiceFactory.getLeAudioService();
            if (service != null) {
                service.setCcidInformation(userUuid, ccid, contextType);
            }
        }
        return ccid;
    }

    /**
     * Release the acquired Ccid
     *
     * @param value Ccid value to release
     */
    public static synchronized void releaseCcid(int value) {
        sAssignedCcidList.remove(value);
        sUserMap.entrySet().removeIf(entry -> entry.getValue().first.equals(value));
    }

    /**
     * Get Ccid information.
     *
     * @return Map of acquired ccids along with the user information.
     */
    public static synchronized Map<ParcelUuid, Pair<Integer, Integer>> getUserCcidMap() {
        return Collections.unmodifiableMap(sUserMap);
    }
}
